[9.0][IMP] Relicense to LGPL-3

lasley opened the pull request:
This is a formal proposal to relicense Vertical Medical to LGPL-3. I will also be pursuing this for v10.

We have breached into website territory in the incubators, and thus customers cannot theme their instances without LGPL. This is an absolute show stopper, and must be resolved in order for this repo to continue further.

I have attempted to gather pre-approval from license holders via private email. 

The follow license holders are non-responsive:

* Tech Receptives
* Acsone (admittedly it's only been two days since I emailed Acsone, but I know they're active here so it was just easier to move to the final forum)

The following license holders have already agreed to the change:

* LasLabs
* Vauxoo

cc @nhomar @acsone @OCA/board @laslabs

elicoidal commented:
:+1:

lmignon commented:
for acsone ping @sbidoul

dreispt commented:
For me LGPL is perfectly OK for technical modules, such as what we can find on "server-tools" and "web".
But for functional modules, that a different story.
If the point is about website themes, why is the AGPL license a problem?
Also, arguably "data modules" are not considered "code", and could coexist with "AGPL source code".

sbidoul commented:
@lasley Sorry for the delay.

In general I'm not in favor of LGPL relicensing, because it makes it too easy for people to create closed source repackaging of our work.

That said our contributions in this repo are comparatevely small. I'm a bit worried of creating a precedent though. 

I'm happy to discuss of course, privately if needed.

lasley commented:
@dreispt @sbidoul - As noted, AGPL does make it easier for people to close source their work. The problem is when this comes into play with website modules, and also the medical industry in general.

In order to create a theme, you must be dependent on the module you are theming. According to Odoo SA - Adding a dependency to your module creates a modified work, therefore it is not possible for a customer to create a theme based on an AGPL website module without having to provide the world their theme. It is my understanding that is why LGPL is now the standard for the website PSC, and Odoo core.

It is a full blocker to require my customers to open source their themes, as I'm sure everyone will agree. Customers spend tens of thousands of dollars in order to theme their website, the last thing they want to hear is that they need to open source their theme for all their competitors to use.

I completely understand the point about setting a precedent, but I am between a rock and a hard place here. As this repo continues to advance, it will continue to add modules related to website. It will also continue to add base modules which may need some proprietary logic in derivatives.

Bringing this out of website context, I'll speak directly on AGPL in the medical arena. This industry is full of proprietary information and industry practices. I cannot open source modules based on these processes, and am currently under NDA for multiple techs & workflows that I plan on developing modules for. I also cannot develop an Open Source module if I am under an NDA for any related aspect of the module. This is all too common in the medical arena and thus makes this EMR fairly useless in all but third world contexts. Let me know if you want some more concrete examples here, I'm only under NDA on the how and not the what.

lasley commented:
> Also, arguably "data modules" are not considered "code",

It is my understanding that Odoo SA has asserted that if a module has another module in the depends section of the manifest (or inherited manifests), the dependant module is a modified work of the original. An attorney in court would need to argue this nuance, and I do not want to be the one funding that TBH.

dreispt commented:
@lasley Licenses apply to source code, and not to data and configurations.
Although packaged in the same way, arguably data only modules are conceptually different from code modules implementing features and logic. About themes, AFAIK Odoo SA have publicly stated that they consider them "data modules". 
But I admit that this can be a grey area.

Regarding the re-licensing, like @sbidoul said, as a general principle AGPL should be used.
But of course exceptions are OK, if good reasons are given.
To allow support for industry specific proprietary information and practices can be an acceptable explanation.
Another good explanation is that you will be the main driving force for this repository development, and without the LGPL you won't be able to open source any code at all, at that wouldn't be good.

PS: <rant> looking from this side of the big river it's quite hard to see, at the management level, how a national health system can function with secret and proprietary information and practices: proper health care requires multiple organizations to work together and relies heavily on commonly known and shared practice protocols. </rant>

lasley commented:
> About themes, AFAIK Odoo SA have publicly stated that they consider them "data modules".

Can you point me to where they say this? I'm honestly having trouble understanding the reasoning behind this statement, because a theme is code. The CSS code in a theme targets elements provided by an AGPL module, and therefore makes changes to it. In some instances, themes can also move elements around the page.

I consider this the exact same thing as inheriting a class and overriding a method. We have been operating under this premise in website PSC, so I'm sure @yajo would also be interested in this.

> it's quite hard to see, at the management level, how a national health system can function with secret and proprietary information and practices

This is the definition of US healthcare in a nutshell, and yeah it pretty much doesn't work. But it is what it is, and this is the market I am in.

My most recent NDA is for electronic pre-authorization through [Cover My Meds](https://www.covermymeds.com). At first glance, it looks like they have themselves a nice open protocol and API. After talking to their account managers, you learn that this open API on their website will be deprecated next year and replaced by a closed API. In order to get those specs, I had to sign an NDA & thus can't open source the module. Any customer deciding to use this module/service must be under an NDA & a BAA with me in order to do so. Creating this feature without them would require me to nail down contracts with thousands of providers. 

I am also under licensing restrictions for the [UMLS Metathesaurus](https://uts.nlm.nih.gov/help/license/LicenseAgreement.pdf), which does not allow redistribution of licensed products except under some specific circumstances. One of these circumstances is that you have to pay for [access to SNOMED from non-member countries](http://www.snomed.org/snomed-ct/get-snomed-ct). 

The list goes on, but the point is that we do not have a shared, free, open, standardized, or functional health care industry in America. We now have Trump, so this is not going in a direction for the better. We're building a wall instead...

yajo commented:
I did not contribute a single LOC here, but now that you ping me, here's my opinion: think: Am I going to sue any AGPL violations of my code? If the answer is no, then I prefer to use LGPL and then legalize the use that I'm not going to persecute anyway. So, I'd +1 to this and forget about it. But I know not everybody thinks this way...

OTOH, I like the Lugaru example. [It's a game that got GPL'd with one of the firsts Humble Bundles, but they applied a different license to the game engine and logics than to assets](https://gitlab.com/osslugaru/lugaru#licensing). [This is perfectly allowed by the [A]GPL for videogames](https://www.gnu.org/licenses/gpl-faq.html#WhatCaseIsOutputGPL), and the assets used to display a theme customized for a customer is exactly the same case: you can have an AGPL theme that has logos, fonts, etc. from a different license, and that's perfectly legal AFAICS.

Not sure if your customers will care about the code that powers their themes (I think people does not care about that mostly), or just by how it looks and their trademarks. I invite all of you to carefully read section 7 of the AGPL. This allows the project to add some additional rules or exceptions to the license itself. [The FSF FAQ recommends an exception for templates](https://www.gnu.org/licenses/gpl-faq.html#WMS), which is mostly OCA's concern about LGPL vs AGPL.

About APIs, I'm not sure but even if you could copyright an API, I don't think you can copyright its implementations. Even if that's the case, you can create a LGPL addon that just speaks to the API and link to it from other AGPL addons.

lasley commented:
>  think: Am I going to sue any AGPL violations of my code?

Can you selectively enforce a license in this way? It was my understanding that as a license holder you have to enforce all terms of the license as written, or a precedence is set that invalidates the license as a whole.

> OTOH, I like the Lugaru example.

In the case of a game, there is a clear isolation of the models from the sprites. I think we're probably fine with CSS under this distinction, but I do not believe this to be the case with QWeb.

Some time back, this question came up in the Wordpress community. [This article](https://wordpress.org/news/2009/07/themes-are-gpl-too/) gives a legal opinion from the Software Freedom Law Center clarifying the distinction between themes and code, which I believe is similarly relevant here. 

The paraphrased relevant portions of the opinion is that the CSS is served directly by the webserver & could be applied to any contexts. The PHP in Wordpress themes, which is used in the same fashion we use Python & QWeb combo, is derived. He even went on to comment about the HTML portion, which is also considered derived.

My thought process in relation to this repo is that none of the data models, methods, or HTML would exist without the currently AGPL-licensed code from this repo. If a module is created that uses QWeb to move an element from the right side of the page to the left side of the page, a modified work of the original is being created and is required to be licensed under compatible terms.

>  The FSF FAQ recommends an exception for templates, which is mostly OCA's concern about LGPL vs AGPL.

This actually might be the ideal solution. We would need to add something for Qweb into their recommended verbiage too, I think. Some of my customers are planning some fairly complex design switch-ups & are paying a lot of money specifically to be unique in these switch ups. The element positioning and all that would definitely play into this.

> I'm not sure but even if you could copyright an API, I don't think you can copyright its implementations

Me neither, but they did & they provided explicit instructions that the module could not be open sourced.

These guys just got purchased by McKesson though, so I'm probably going to kick that plan to the curb anyways simply out of dislike for the parent company. IMO they're one of the main problems in the US & are behind a lot of consolidations that are causing some bad impacts in healthcare.

elicoidal commented:
I  have the feeling that we are restarting here the whole AGPL/LGPL discussion that happened time ago between the board and Odoo when Odoo decided to switch licenses.
I do not really see the point to restart the whole discussion for this particular case knowing that a consensus has been reached with Odoo time ago.
I am probably not the most suitable guy here to give all details but basically the LGPL has been selected by Odoo to allow protection for web designers (same case as @lasley is pointing out now).

Please keep in mind that the OCA has nothing against the LGPL per se. OCA will keep the code open whatever other contributors might do. If the current contributors to the code do agree to switch, I do not see any reason not to.

Do not forget that if some new contributors want to close his contribution and the code stays AGPL, he will simply not contribute and the project will not bloom. Going LGPL in this particular case will help the universe to grow and ensure that that the part of the code in hand of the OCA is well maintained as the community grow.

lasley commented:
> Do not forget that if some new contributors want to close his contribution and the code stays AGPL, he will simply not contribute and the project will not bloom. 

This is the problem I am in right now. TBH I hate to place an ultimatum, but my current standpoint is that this needs to go LGPL or I will be moving on from this repo. Seeing as I'm basically the only maintainer, this would be a crushing blow to vertical-medical, but I simply cannot ask my customers to open source their design changes.

lmignon commented:
@lasley I'm a little bit afraid by the change of licence. If this vertical repository becomes LGPL it's no more possible to depend of AGPL modules. Do you plan to ask for a licence change each time you'll need to include an other AGPL module from OCA. In such a case, this change should silently contaminate all the OCA ecosystem...

lasley commented:
@lmignon - I believe you misunderstand the implications of LGPL in this context. 

Any derivative or modified work of LGPL-3 is compatible for licensing under:
* LGPL-3
* GPL-3
* AGPL-3

Any modified or derivative work of AGPL-3 is compatible for licensing under:
* AGPL-3

Compatible licenses can be chosen by the author at their whim. My issue here is that our only compatible license is AGPL-3 for this entire repo, which then umbrellas the website side of things. 

In order to depend on an AGPL module, you only need license the current module you're working in as AGPL, and make sure that any modules that depend on the current one are also AGPL. This strategy is already confirmed by Odoo SA, with their licensing of Odoo as LGPL & the fact that there has been no silent contamination of AGPL modules since that change.
 
I created LasLabs#129 detailing the implications of the license change for the unreleased version of Vertical Medical, which does include a few AGPL assets as dependencies. The v9 version contains no external dependencies & thus no implications.

I actually still plan on releasing a lot of AGPL for this repo (importers, etc), but I need to be able to isolate accordingly. As it stands now, it's all AGPL-3 or bust because the core data models are AGPL-3.

dreispt commented:
I am convinced that the relicensing here is the best move here.
It should be underlined that AGPL is the general rule, and this is and exception.
@lasley has been doing very significant contributions here, and the LGPL would allow him to continue doing so.
Not being to continue would be a greater loss for this vertical and the OCA.

JordiBForgeFlow commented:
Laurent and Dave,

In this case the medical project would not be able to build modules that
depend on AGPL modules.

But anyone would be able to build an AGPL module that depends on the
LGPL'ed vertical modules.

LGPL is, therefore, more permissive with respect what can you do with
derivative works.

I would say that the project work has been authored by Laslabs by far, and
it would be fair to respect his will to choose the license that best fits
his needs.

Now, if the project starts to require features that can be found in AGPL'ed
modules from other OCA repos, and the respective authors are not willing to
change their license, the vertical-medical will struggle. But so far
Laslabs has been able to either convince others to change their license or
to walk around the AGPL'ed modules.

The discussion of LGPL vs. AGPL is, to me, equivalent to capitalimsm vs.
communism. And very related to the differences between open source software
and free software. LGPL works best for , I'd say, pretty large companies,
where the big money is, because it protects the private where needed. Open
source sofware is for companies a growth strategy, but not a show stopper
towards licensing/private path. AGPL works best when it comes to protect a
community that wishes to grow from shared participation, and towards
enrusing the continuity free software. So, same code, two radically
different ways to perceive the world.

I would think that in general, if various companies compete using core LGPL
licensed code + private add-ons, and the number of companies and codebase
is sufficiently large, some companies will choose to LGPL features that
others chose to close. If you overlap the LGPL'ed features, you could end
up with a pretty significant and usable LGPL'ed codebase.

For example, Laslabs publishes their vertical-medical under LGPL and
provides a closed-source solution for reporting that builds upon the
vertical-medical.
On the other side, Acme Inc. publishes a LGPL'ed reporting that builds also
on the vertical-medical, and chooses to sell a closed-source solution for
medical billing.

In the end, using the combined Laslabs + Acme Inc. LGPL'ed code you could
end up with a medical solution that includes a decent reporting, all under
LGPL, and still Laslabs and Acme could do business by selling their private
code to a fraction of companies that prefere Laslab's reporting or Acme's
billing over the LGPL'ed version, or other competitor's solutions.

Regards,
Jordi.






Regards,
Jordi.

On Mon, Feb 6, 2017 at 5:37 PM, Laurent Mignon (ACSONE) <
notifications@github.com> wrote:

> @lasley <https://github.com/lasley> I'm a little bit afraid by the change
> of licence. If this vertical repository becomes LGPL it's no more possible
> to depend of AGPL modules. Do you plan to ask for a licence change each
> time you'll need to include an other AGPL module from OCA. In such a case,
> this change should silently contaminate all the OCA ecosystem...
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/vertical-medical/pull/166#issuecomment-277737483>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHU_VuhajW2MQW9feqsBthQGejEcoe-Gks5rZ0xVgaJpZM4LvaTl>
> .
>



-- 
Jordi Ballester Alomar
CEO & Founder | Eficent
(+34) 629530707 | jordi.ballester@eficent.com | http://www.eficent.com
Twitter: https://twitter.com/jbeficent_erp | Linkedin:
https://www.linkedin.com/in/jordiballesteralomar

lasley commented:
Thank you all for your input. 

Does anyone have additional input to provide, or an explicit approval/disapproval? I would like to move vertical-medical tickets into our next sprint, and I can't do that until I know what the plan is here.

lasley commented:
I am particularly curious to hear from @parthivgls. Sorry I didn't tag you originally, I didn't know your Github until @dreispt tagged you!

pedrobaeza commented:
I think you can move on...

sbidoul commented:
@lasley Sorry I lost sight of this one. 

As you understood I'm too fond of re licensing to LGPL. As @jbeficent says, this is some sort of political statement indeed, but IMHO not on the communism vs capitalism axis. It's rather a matter of guaranteeing user/customer freedom, foster open architectures and avoid vendor lock-in. I think that open source in the area of business applications is still a fragile endeavor that needs help from copyleft licenses.

Now you say you will stop contributing here if it stays AGPL. Is that not cheaper for you to devise an architecture that isolates proprietary bits in loosely coupled components than rewriting all vertical-medical? 

That said, Acsone is not a major contributor of this repo, so we won't block the decision of other authors.

CLA grants the right to re license but current OCA policy is to obtain agreement from authors before doing it, so please make sure all authors of non-trivial contributions to vertical-medical agree before merging.

lasley commented:
Thanks for the response @sbidoul 

> Is that not cheaper for you to devise an architecture that isolates proprietary bits in loosely coupled components than rewriting all vertical-medical?

I have tried to devise this, but with the core data models as AGPL I cannot. For the most part, anything I do is at least coupled with the data models.

I can look into making more of vertical-medical AGPL instead of changing everything to LGPL like I am here, but without the core data models I cannot make proprietary models, reports, or views.

Some vendors have specifically told me that even their data model is proprietary, so I must close source any module I make with any trace of those schemas. I don't fully understand the legality here, but the fact of the matter is that they have specifically told me no & legal or not, they can simply stop working with me. 

In this sense, it is significantly cheaper for me to recreate everything than to lock myself from being able to make or adhere to vendor agreements. It sucks, particularly with how much I've invested here, but it is what it is unfortunately. 

> CLA grants the right to re license but current OCA policy is to obtain agreement from authors before doing it, so please make sure all authors of non-trivial contributions to vertical-medical agree before merging.

This is why I am really trying to not just press forward here. I fully respect this rule, and do not want to stomp on toes. I'm not even cursing at the need to recreate TBH, I simply need to know.

The remaining contributor here is Tech Receptives, who have not responded to email request & now pings on this matter. They started this repo, so this is a serious issue to press forward without them IMO. I'm not really sure how else to get their attention though.

parthivgls reviewed (approved):
I am fine with the license change as far as it helps make this project progress.

lasley committed:
[IMP] Relicense to LGPL-3

lasley commented:
Hooray thanks @parthivgls! 

Final tally:

Approve:
* Tech Receptives
* Vauxoo
* LasLabs

Abstain:
* Acsone

I have rebased and re-readied this PR for merge. GNU now supplies an [official markdown version](https://www.gnu.org/licenses/lgpl-3.0.md) of their licenses, so I switched the plaintext AGPL to a markdown LGPL too

elicoidal commented:
@dreispt 
> It should be underlined that AGPL is the general rule, and this is and exception.

The rule of OCA is any OSI-compatible license not only AGPL (https://opensource.org/licenses).

sisalp commented:
So, after a first code parsing, it seems to me this code had more contributors than said and original names have been probably intentionally wiped off the code, but not completely.

I don't know which part of the original work is still in use today, but at least the authors of the original openerp-medical project are still mentioned in some parts of the code.

This vertical has been known for multiples hacks and property violations since it moved outside OpenERP. I remember blocking on my servers, some years ago, a medical module which was an illegal copy of this original project. This module was sold on apps, the case was reported to Odoo with no action and this violation was made public. The code we have today may b, at least partially, an evolution of this code.

May I recommend you to be very careful with this matter since, if I understand correctly recent talks, the original authors have never been asked when their code was moved to OCA, and they never signed the CLA.
I hope this will end in a less important issue than it seems to me at first.

sisalp commented:
For making your study, I found the origimal medical work, I recorded in 2011.
http://download.sisalp.net/medical_1_0_0.tar.gz
Compairing this work with https://launchpad.net/oemedical I'm afraid of understanding Tiny sprl was the original barbarian

the first one :
#    Copyright (C) 2008-today  Luis Falcon

the derivated :
# Copyright (C) 2004-2009 Tiny SPRL (<http://tiny.be>). All Rights Reserved

 Hope I'm mistaken.

lasley commented:
Clear fixes in #187 & #188 for v10 and v9 respectively

lasley commented:
#187 & #188 submitted to remove legacy modules from new branches & update the main project description.

In my capacity as PSC lead for Vertical Medical, I have decided that the safest option for v7 and v8 is to delete them. In this light, I have removed those branches from OCA & have requested that Odoo remove them from the Apps store.

The situation is not fully resolved at this time, as a code audit still needs to take place for the existing code. I am fairly confident, however, that the current code is correct in regards to attribution & doesn't maintain a vague resemblance to the code in question.

As mentioned in my pull requests - Thymbra is welcome to claim specific files as well, as any standard copyright/DMCA process would require. I will personally respond and rectify whatever the issue is immediately when identified.

We would prefer these claims be submitted to legal@odoo-community.org, but I will also be watching Github. I do not feel that Twitter is an appropriate avenue to resolve this situation, however.
